localstack/services/s3/provider_native.py

```diff
@@ -1259,6 +1289,510 @@ def restore_object(
+
+        return response
+
+    def complete_multipart_upload(
+        self,
+        context: RequestContext,
+        bucket: BucketName,
+        key: ObjectKey,
+        upload_id: MultipartUploadId,
+        multipart_upload: CompletedMultipartUpload = None,
+        checksum_crc32: ChecksumCRC32 = None,
+        checksum_crc32_c: ChecksumCRC32C = None,
+        checksum_sha1: ChecksumSHA1 = None,
+        checksum_sha256: ChecksumSHA256 = None,
+        request_payer: RequestPayer = None,
+        expected_bucket_owner: AccountId = None,
+        sse_customer_algorithm: SSECustomerAlgorithm = None,
+        sse_customer_key: SSECustomerKey = None,
+        sse_customer_key_md5: SSECustomerKeyMD5 = None,
+    ) -> CompleteMultipartUploadOutput:
+
+        store = self.get_store(context.account_id, context.region)
+        if not (s3_bucket := store.buckets.get(bucket)):
+            raise NoSuchBucket("The specified bucket does not exist", BucketName=bucket)
+
+        if not (s3_multipart := s3_bucket.multiparts.get(upload_id)):
+            raise NoSuchUpload(
+                "The specified upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.",
+                UploadId=upload_id,
+            )
+
+        # TODO: validate key?
+        if s3_multipart.object.key != key:
+            pass
+
+        parts = multipart_upload.get("Parts", [])
+        if not parts:
+            raise InvalidRequest("You must specify at least one part")
+
+        parts_numbers = [part.get("PartNumber") for part in parts]
+        # sorted is very fast (fastest) if the list is already sorted, which should be the case
+        if sorted(parts_numbers) != parts_numbers:
+            raise InvalidPartOrder(
+                "The list of parts was not in ascending order. Parts must be ordered by part number.",
+                UploadId=upload_id,
+            )
+
+        parts_fileobjs = self._storage_backend.get_list_parts_fileobjs(bucket, upload_id)
+        version_id = generate_version_id(s3_bucket.versioning_status)
+        dest_key_fileobj = self._storage_backend.get_key_fileobj(bucket, key, version_id)
+
+        try:
+            s3_multipart.complete_multipart(parts, dest_key_fileobj, parts_fileobjs)
+        except Exception:
+            self._storage_backend.delete_key_fileobj(bucket, key, version_id)
+            raise
+
+        self._storage_backend.delete_multipart_fileobjs(bucket, upload_id)
+
+        s3_object = s3_multipart.object
+
+        if s3_bucket.versioning_status == "Enabled" and (
+            existing_s3_object := s3_bucket.objects.get(key)
+        ):
+            existing_s3_object.is_current = False
+
+        s3_bucket.objects.set(key, s3_object)
+
+        # remove the multipart now that it's complete
+        self._storage_backend.delete_upload_directory(bucket, s3_multipart.id)
+        s3_bucket.multiparts.pop(s3_multipart.id, None)
```

Comment:
Same as above, I liked the orchestration behind done explicitly in the provider, so that metadata and assets would be separated. Right now, the StorageBackend has one job, and it's giving away file objects for different kind of usage. I think it seems a bit empty right now because it only does the fake filesystem, but the one for persistence would have more substance to it and have abstraction around filesystem directories, listing and retrieving files inside of them, and mixing S3 logic in that felt a bit weird?